### PR TITLE
[TechDocs] Local Publisher Case Sensitivity Compatibility

### LINKED ISSUE #6979: TechDocs returning "ERROR 404: i: Page not found"
## Expected Behavior

TechDocs should generate docs properly when opening a docs component from the main menu.

## Current Behavior

When opening a Docs site that is configured to use either github or azure devops, you get a 404 error page suggesting that perhaps you are missing an index.md but if you look at the browser console there is a 404 error loading an index.html from an API route. However, there seems to be an issue with the URL itself. Once I change the part of the URL from lower to upper case ("component" to "Component"), everything works fine. Also, when opening techdocs directly from the catalog (https://backstage.xxx.com/catalog/default/component/xxx/docs), it works fine.

## Possible Solution

Currently, the following URL is opened after clicking on the techdoc:
`https://backstage.xxx.com/docs/default/component/xxx`

It works after changing the "Component" to upper case:
`https://backstage.xxx.com/docs/default/Component/xxx`

## Steps to Reproduce

1. Generate app using @backstage/create-app@0.3.38
2. Integrate component with a techdoc from Azure or Github
3. Open integrated component from the main menu (Docs -> component name)

## Context

Error text displayed in UI:
```
ERROR 404: i: Page not found. This could be because there is no index.md file in the root of the docs directory of this repository.

Looks like someone dropped the mic!
Go back... or please contact support if you think this is a bug.
```

## Your Environment

- Docker image based on node:14-buster-slim
- App generated using: @backstage/create-app@0.3.38
- Edge / Chrome / Firefox

### LINKED ISSUE #6980: docs are not available from UI latest backstage version
1. I have component definition like
```
apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  name: my-service
  description: |
    Main brain
  tags:
    - kotlin
  links:
    - title: Playbook entry
      url: https://domain.com/playbook/services/my-service.html
  annotations:
    backstage.io/source-location: url:https://gitlab.domain.com/xxx/my-service
    backstage.io/techdocs-ref: url:https://gitlab.domain.com/xxx/my-backstage/-/tree/master/my-service
spec:
  type: service
  owner: myteam
  lifecycle: production
  dependsOn: ['component:talon-one']
  providesApis: ['my-service-api']
```
2. Then backstage parses docs and displays them on the docs index page
3. When I click on the record, it says "dropped the mic!" and I see that it tries to query https://domain.com/api/techdocs/static/docs/default/component/my-service/index.html (returns 404)
4. I see that https://domain.com/api/techdocs/static/docs/default/Component/my-service/index.html returns content

## Expected Behavior

Documentation is shown

## Current Behavior

Error is shown

## Possible Solution

When I login into the container I can manually rename folder to lowercase and it starts to show it:
```
root@backstage-75cf748684-7rl9n:/# cd /app/node_modules/@backstage/plugin-techdocs-backend/static/docs/default/
root@backstage-75cf748684-7rl9n:/app/node_modules/@backstage/plugin-techdocs-backend/static/docs/default# ls
Component
root@backstage-75cf748684-7rl9n:/app/node_modules/@backstage/plugin-techdocs-backend/static/docs/default# mv Component component
```

## Context

I've updated backstage with `yarn backstage-cli versions:bump` and did all necessary patch changes and now documentation is not working.
Another thing which is related to this issue is that [DefaultTechDocsCollator](https://github.com/backstage/backstage/blob/master/plugins/techdocs-backend/src/search/DefaultTechDocsCollator.ts) queries `/api/techdocs/static/docs/default/Component/my-service/search/search_index.json` (capitalised "Kind"). So I can not just rename that folder, because then collator does not work.

It is also worth to mention, that `DefaultTechDocsCollator` queries an `url` with capitalised `C`:
`/api/techdocs/static/docs/default/Component/my-service/search/search_index.json`. So there is currently some misalignment. Maybe both variants shall be served, from the back-end perspective.

## Your Environment
Docker container build on top of `FROM node:14-buster-slim`

### PR DESCRIPTION
**Pull request by iamEAP**
## Hey, I just made a Pull Request!

In #6709, we made the assumption that "no changes are necessary to the local publisher because if the docs aren't found, they'll just be regenerated with the correct case."  We also ran into no issues testing this assumption because MacOS' file system is case-insensitive by default and so nothing seemed awry when we were testing locally.

This corrects things by bringing the `Local` publisher up-to-snuff with the GCS/AWS/Azure publishers by...

- [x] Lower-casing entity triplets on write/publish.
- [x] Lower-casing entity triplets on simple reads (`hasDocsBeenGenerated` and `fetchTechDocsMetadata`)
- [x] Redirects requests for non-lowercase entity triplet objects to their lowercase equivalents (there is otherwise no easy way to continue to take advantage of the `express.static()` middleware)
- [x] Respecting the `techdocs.legacyUseCaseSensitiveTripletPaths` (by undoing all of the above when set to `true`)

Closes #6979 #6980 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Tests for new functionality and regression tests for bug fixes
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/techdocs-common/src/stages/publish/local.ts [hunk 1]
```diff
@@ -81,8 +87,7 @@ export class LocalPublish implements PublisherBase {
   publish({ entity, directory }: PublishRequest): Promise<PublishResponse> {
     const entityNamespace = entity.metadata.namespace ?? 'default';
 
-    const publishDir = path.join(
-      staticDocsDir,
+    const publishDir = this.staticEntityPathJoin(
```
Comment (camilaibs) [inline packages/techdocs-common/src/stages/publish/local.ts:90]: Should we check if the legacy case flag is not configured before?

**Review by camilaibs — APPROVED** [review]

**Comment by camilaibs** [comment]
@iamEAP do we need to fix it here too?

https://github.com/backstage/backstage/blob/techdocs/beta-local/plugins/techdocs-backend/src/search/DefaultTechDocsCollator.ts#L82

**Comment by iamEAP** [comment]
> _do we need to fix it here too?_
> 
> _https://github.com/backstage/backstage/blob/techdocs/beta-local/plugins/techdocs-backend/src/search/DefaultTechDocsCollator.ts#L82_

Good thought @camilaibs!  I was thinking that since we don't currently check config at all in there , it's probably best to keep it as-is for now.

The thinking is:
- As things are by default, loading `/namespace/Kind/name/search/search_index.json` should work just as well as `/namespace/kind/name/search/search_index.json` (and the 301 redirect in this PR achieves that for local).
- When the legacy config is set, _only_ `/namespace/Kind/name/search/search_index.json` will load.

Which means the current way of loading it will work regardless of the flag, while changing it would mean it only works when the flag is not set.

After this has been out and stable a little while longer, we can clean it up.

**Comment by camilaibs** [comment]
> > _do we need to fix it here too?_
> > _https://github.com/backstage/backstage/blob/techdocs/beta-local/plugins/techdocs-backend/src/search/DefaultTechDocsCollator.ts#L82_
> 
> Good thought @camilaibs! I was thinking that since we don't currently check config at all in there , it's probably best to keep it as-is for now.
> 
> The thinking is:
> 
> * As things are by default, loading `/namespace/Kind/name/search/search_index.json` should work just as well as `/namespace/kind/name/search/search_index.json` (and the 301 redirect in this PR achieves that for local).
> * When the legacy config is set, _only_ `/namespace/Kind/name/search/search_index.json` will load.
> 
> Which means the current way of loading it will work regardless of the flag, while changing it would mean it only works when the flag is not set.
> 
> After this has been out and stable a little while longer, we can clean it up.

I'm seeing this warning when the server starts:
image

Does it have a relation?

**Comment by iamEAP** [comment]
> I'm seeing this warning when the server starts:
> image
> 
> Does it have a relation?

Good question!  I _think_ this is unrelated.  This warning will appear any time the TechDocs collator tries to index a site that has not been built yet (I see this from time to time when I wipe out local directories, for example).

We should probably change the log level for this, though.  It's probably not worth warning on and we want to reduce log fatigue.

**Pull request merged**